apps/forms/avoid-losing-form-data/src/app/ui/form.component.ts

```diff
@@ -63,16 +71,33 @@ import { FormBuilder, ReactiveFormsModule, Validators } from '@angular/forms';
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
 export class FormComponent {
-  private fb = inject(FormBuilder);
+  hasValues = output<boolean>();
 
-  form = this.fb.nonNullable.group({
+  #fb = inject(FormBuilder);
+
+  form = this.#fb.nonNullable.group({
     name: ['', { validators: [Validators.required] }],
     email: ['', [Validators.required, Validators.email]], // other syntax
     phone: '',
     message: '',
   });
 
+  constructor() {
+    this.form.statusChanges
+      .pipe(
+        map((form) => Object.values(form)),
+        map((values) => values.map((v) => v.trim()).some((v) => v.length)),
+        distinctUntilChanged(),
+        tap((hasValues) => this.hasValues.emit(hasValues)),
+        takeUntilDestroyed(),
+      )
+      .subscribe();
```

Comment:
why don't you listen to form.dirty, form.touched and add a submit flag ? 
it seems easier to understand 😅
but your solution works as well. But you will have to write it in all your form components.

Reply:
Because a form can still be marked as dirty even if the user starts typing and manually cleans the form afterward, if I remember correctly?

"But you will have to write it in all your form components." maybe i could create an RxJS operator for that